src/lib/lore.sh

```diff
@@ -411,112 +411,77 @@ function get_patches_from_mailing_list()
-
-  # Replace space in favor of _
-  file_name="${title// /_}"
-
-  # Replace .
-  file_name="${file_name//./_}"
-
-  # Replace special character
-  file_name="${file_name//[&*+%!:,]/}"
-
-  # Replace /
-  file_name="${file_name//[\/\\]/-}"
-
-  printf '%s' "$file_name"
-}
-
-function convert_title_to_patch_name()
-{
-  local title="$1"
-  local index="$2"
-  local file_name
-
-  file_name=$(title_to_path "$title")
-
-  if [[ -n "$index" ]]; then
-    printf '%04d-%s\n' "$index" "${file_name}.mbox"
-    return
-  fi
-
-  printf '%s\n' "${file_name}.mbox"
-}
-
-function url_update_patch_number()
-{
-  local url="$1"
-  local new_number="$2"
-
-  url="${url/-[0-9]*-/-$link_ref-}"
-
-  printf '%s' "$url"
-}
-
+# This function downloads a patch series in a .mbx format to a given directory
+# using a series URL. The series URL should be the concatenation of the lore URL, the
+# target mailing list, and the message ID. Below is an example of such a URL:
+#   https://lore.kernel.org/some-list/2367462342.4535535-1-email@email.com/
+# The output filename is '<message ID>.mbx'. This function uses the `b4` tool underneath
+# to delegate the process of fetching and downloading the series. The file generated is
+# ready to be applied into a git tree, containing just the patches in the right order,
+# without the cover letter.
+#
+# Return:
+# Return 0 if the thread was successfully downloaded, 22 if the series URL or the output
+# directory passed as arguments is empty and the error code of `b4` in case it fails.
 function download_series()
 {
-  local total_patches="$1"
-  local first_message_id="$2"
-  local save_to="$3"
-  local title="$4"
-  local total=0
-  local link_ref=0
+  local series_url="$1"
+  local save_to="$2"
+  local flag=${3:-'SILENT'}
   local url
-  local patch_file_name
+  local series_filename
+  local cmd
+  local ret
 
-  mkdir -p "$save_to"
+  # Safety checking
+  if [[ -z "$series_url" || -z "$save_to" ]]; then
+    return 22 # EINVAL
+  fi
 
-  url=$(replace_http_by_https "$first_message_id")
+  # Create the output directory if it doesn't exists
+  cmd_manager "$flag" "mkdir --parents '${save_to}'"
+  ret="$?"
+  [[ "$ret" != 0 ]] && return "$ret"
 
-  until ! is_the_link_valid "$url"; do
-    ((total++))
-    ((link_ref++))
+  # Although, by default, b4 uses the message-ID as the output name, we should assure it
+  series_filename=$(printf '%s' "$series_url" | rev | cut --delimiter '/' -f2 | rev)
 
-    url=$(url_update_patch_number "$url" "$link_ref")
-    patch_file_name=$(convert_title_to_patch_name "$title" "$link_ref")
-    download "${url}raw" "$patch_file_name" "${save_to}" &
-  done
-  wait
+  # For safe-keeping, check the protocol used
+  url=$(replace_http_by_https "$series_url")
+
+  # Issue the command to download the series
+  cmd="b4 --quiet am '${url}' --no-cover --outdir '${save_to}' --mbox-name '${series_filename}.mbx'"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  [[ "$ret" != 0 ]] && return "$ret"
```

Comment:
Do you know which type of error b4 can return? If so, we should consider a better return log with a reasonable error message.

Reply:
Nice catch.
Could reproduce 2 kinds of errors returned by `b4` with this template of execution:

- Invalid arguments returns 2 ENOENT (strange, should return 22 EINVAL)
- Unknown URL (message-ID) and invalid `save_to` returns 1 EPERM. Also strange. Why the same errno, if the error messages displayed are different?

Either way, two types of errors are better than just a generic one 😆